MRG, ENH:  Use peak times for sensor and source plots in .html reports

NeuroLaunch opened the pull request:
A few comments:
- As @larsoner suggested, I used the global field power from the evoked object; for the source section, if that was already calculated for the sensor section, then I reused the 'times' for that particular analysis file / condition(name) combination (i.e. memoization)
- I went with the usual standard-deviation approach for gfp, but sqrt of the mean of the square is another way to go that takes into account mean activity across channels; I don't think the differences were that large, especially for the largest peaks
- There are other peak-finding functions, such as _find_peaks (and _process_times) in viz.utils; however, I think the prominence approach is more robust
- Per _process_times(), I assumed a modest number of peaks across the gfp, taking the most "prominent" and maintaining temporal order; however, peaks at negative event times are possible (but might still be of interest for a report because they give a sort of baseline)
- I tested this on the funloc subject #1; if there are alternative test data sets, please let me know and I can try them
- In line 503 or so of ._report.py in master, 'times' for the SNR report is defined but not used.  I didn't change this myself, though
- Finally, I'm not sure where to put tags such as [ci skip], or if that's even appropriate here.

larsoner commented on mnefun/_report.py:
```diff
@@ -251,6 +252,37 @@ def _report_raw_psd(report, raw, raw_pca=None, p=None):
+    if 'meg' in evoked:
+        evk = evoked.copy().pick_types(meg=True)
+        gfp = evk.data.std(axis=0)
+        gfp_list.append(gfp/gfp.max())
```
by local convention we follow the PEP8 spacing between separators
```suggestion
        gfp_list.append(gfp / gfp.max())
```

larsoner replied:
... and this is why Travis is unhappy:

https://travis-ci.org/github/LABSN/mnefun/builds/665584883#L393-L406

NeuroLaunch commented on mnefun/_report.py:
```diff
@@ -251,6 +252,37 @@ def _report_raw_psd(report, raw, raw_pca=None, p=None):
+        times = gfp.argmax()
+    times = evoked.times[times]  # convert units to seconds
+    
+    print('Local peak times calculated at', times, 'sec.')
```
That wasn't intended for printing to the .html, only to the console (like the other print statements (e.g. `print('    %s skipped' % section)`), but I'll try it.

larsoner replied:
Yeah it won't end up in the HTML, but if you run the report generation you should see it nicely formatted with the operation and timing, and with this `print` statement in there it will mess it up

larsoner commented on examples/funloc/funloc_params.yml:
```diff
@@ -98,4 +98,4 @@ report:
   sensor: {analysis: All, name: All, times: [0.1, 0.2]}
   source: {analysis: All, name: All, inv: '%s-50-sss-meg-eeg-free-inv.fif',
-           times: [0.09, 0.4], views: lat, size: [800, 400]}
+           times: 'peaks', views: lat, size: [800, 400]}
```
It would be good to also use `"times"` for the `sensor` category above for consistency

NeuroLaunch replied:
Do you mean `times='peaks'` for the sensor category?  I figured showing `times` set in both modes (an array and a string) best shows the available options to a new user.  Alternatively, detailing all options for the various report parameters is something that perhaps should be added to the ever-improving Overview doc.

larsoner replied:
I'd rather leave the detail-all-options in the docs section, the example works better for showing good, consistent option usage

NeuroLaunch commented on mnefun/_report.py:
```diff
@@ -251,6 +252,36 @@ def _report_raw_psd(report, raw, raw_pca=None, p=None):
+        gfp = evoked.data.std(axis=0)
+        gfp = gfp / gfp.max()
+    else:
+        gfp = np.array(gfp_list).mean(axis=0)
```
I think the negative peak times is a result of the "prominence" algorithm, which can skip minor peaks too close to a major one; thus spurious energy at negative times makes the top-5 peaks.  This would likely occur in any scenario where negative times aren't suppressed.  One way around it is to limit the number of peaks (e.g. 3 instead of the current max of 5).  As I wrote before though, at least for the report negative peaks may not be so bad as they provide indication of activity during the baseline period.

If we do definitely want to avoid negative times, without explicitly disallowing them, whitening seems like a good idea.  The example you show above is compelling, as the baseline variance is quite low.  Will there be difficulties when an evoked object doesn't represent continuous time, but peaks are still meaningful for a report (e.g. a spectrum)?

larsoner replied:
> Will there be difficulties when an evoked object doesn't represent continuous time, but peaks are still meaningful for a report (e.g. a spectrum)?

The whitening is spatial so if the spectrum linearity is preserved (i.e., kept complex / not taking the magnitude) it would still be fine. But I think we're a ways away from doing anything like this in the built-in report generation, it seems better suited to custom paradigm-specific code.

Want to take a stab at the whitening to see if it makes things work better?

NeuroLaunch replied:
Yes, I'll be happy to try a whitening approach for time-series peak detection.  It will take a bit of time to make sure I understand all the steps - I'll start as you suggest with the whitening section of the report.

On a side note I was thinking that it would be too restrictive to simply prohibit detection of negative-time peaks.  I can imagine an experimental paradigm for which signal energy prior to some 0-time reference might be expected (e.g. a sequence of event types, like a saccade or response to an auditory cue preceding a flashed visual target).

NeuroLaunch replied:
Ah, and now I see why you said the whitening also normalizes across sensor types: From the figure, the baseline power is 1.0 on average (red dotted line).

larsoner replied:
Agreed it's better not to force non-negative

NeuroLaunch commented:
The new whitened gfp approach seems to work OK.  I used the definition of gfp from mne.viz.evoked.py (rather than std() used in other places in mnefun) for consistency with the mne function _plot_evoked_white().  The one drawback is that if noise_covariance can't be found for whitening, the gfp definition will include a component related to the non-zero mean.

For more robust rejection of negative-time peaks, I set a gfp threshold of 1.0.  We could consider defaulting to  3 peaks instead of 5.

I get an error when the mnefun parameters are set for `mf_autobad_type = 'python'`.  So I've been running my tests for this PR with `mf_autobad = False`.  I can create a new issue for it.

I meant to add @verbose decorator to quell some of the complaints with the rank calculation (as is done in the call to mne evoked.plot_white() within the Whitening report section).  The rank calculations seem to work ok despite the warnings.  Is it worth adding a verbose='error' feature to run _peak_times() without the warnings being logged?

larsoner commented:
> For more robust rejection of negative-time peaks, I set a gfp threshold of 1.0. We could consider defaulting to 3 peaks instead of 5.

Great, that makes sense as a lower bound because of how the whitening works anyway.

> I get an error when the mnefun parameters are set for mf_autobad_type = 'python'. So I've been running my tests for this PR with mf_autobad = False. I can create a new issue for it.

Make sure you are on latest `mnefun` and `mne` master first. The API changed right before 0.20 came out Friday and both were updated to (hopefully) accommodate the change.

> Is it worth adding a verbose='error' feature to run _peak_times() without the warnings being logged?

Yes you can pass `verbose='error'` to the MNE calls. We shouldn't need an `verbose` decorators locally to squash the warning messages.

You want to push those changes then also rebase (or merge your branch with `upstream/master`) and push? There is currently a merge conflict. If you're not sure / can't figure out how to resolve it, I can give some starter commands for how I do it

NeuroLaunch commented:
Yes, I'll try to resolve the merge conflicts and re-push, hopefully by tomorrow.  I did a recent MNE pull, but I'll double-check that I have the latest first.  As for the warning messages, I used a key MNE function directly that didn't have a verbose parameter.  I'll see if there's an easy work-around.

larsoner commented:
You can use the context manager:
```
mne.utils.use_log_level('error'):
    ...
```
Fun fact, internally it's what the `verbose` decorator actually does anyway!

NeuroLaunch commented:
Rebased with latest MNEFun master, added the `use_log_level()` trick, and tested again with updated MNE.  Works, and no conflicts this time with the mf_autobad set to 'python'.

NeuroLaunch commented:
There's a "Close Pull Request" button at the bottom of this comment field.  Shall I press that to remove this Pull Request, given that it's been merged?  (I was surprised to see this still active, as I figured PRs go away on their own after a merge.)

larsoner commented:
I have no idea why that's happening, probably some GitHub bug
